compiler/AST/ForallStmt.cpp

```diff
@@ -0,0 +1,204 @@
+  for_alist(expr, fIterExprs)
+    _this->fIterExprs.insertAtTail(COPY_INT(expr));
+  for_alist(expr, fIntentVars)
+    _this->fIntentVars.insertAtTail(COPY_INT(expr));
+
+  return _this;
+}
+
+void ForallStmt::replaceChild(Expr* oldAst, Expr* newAst) {
+  if (oldAst == fLoopBody) {
+    if (!newAst)
+      fLoopBody = NULL;
+    else if (BlockStmt* newBlock = toBlockStmt(newAst))
+      fLoopBody = newBlock;
+    else
+      // It is caller responsibility to make newAst fit.
+      INT_ASSERT(false);
+
+  } else if (fWith->replaceChildFI(oldAst, newAst))
+    ; //good
+
+  else
+    // We did not find oldAst in our ForallStmt.
+    INT_ASSERT(false);
+}
+
+// Todo: are these checks done elsewhere?
+static void verifyList(AList& list, Expr* parent) {
+  INT_ASSERT(list.parent == parent);
+  for_alist(expr, list) {
+    INT_ASSERT(expr->list == &list);
+    INT_ASSERT(expr->parentExpr == parent);
+  }
+}
+
+void ForallStmt::verify() {
+  Expr::verify(E_ForallStmt);
+
+  INT_ASSERT(fIterVars.length == fIterExprs.length);
+  if (fZippered) INT_ASSERT(fIterVars.length > 0);
+  else           INT_ASSERT(fIterVars.length == 1);
+
+  verifyList(fIterVars, this);
+  verifyList(fIterExprs, this);
+  verifyList(fIntentVars, this);
+  for_alist(expr, fIterVars)
+    INT_ASSERT(isDefExpr(expr));
+  for_alist(expr, fIntentVars)
+    INT_ASSERT(isDefExpr(expr));
+
+  INT_ASSERT(fLoopBody);
+  verifyParent(fLoopBody);
+  verifyNotOnList(fLoopBody);
+  // should be a normal block
+  INT_ASSERT(fLoopBody->blockTag == BLOCK_NORMAL);
+  INT_ASSERT(!fLoopBody->blockInfoGet());
+  INT_ASSERT(!fLoopBody->isLoopStmt());
+  INT_ASSERT(!fLoopBody->useList);
+  INT_ASSERT(!fLoopBody->userLabel);
+  INT_ASSERT(!fLoopBody->byrefVars);
+  INT_ASSERT(!fLoopBody->forallIntents);
+
+  INT_ASSERT(fWith);
+
+  // Currently ForallStmt are gone during resolve().
+  INT_ASSERT(!resolved);
+}
+
+void ForallStmt::accept(AstVisitor* visitor) {
+/* TODO once AstVisitor is updated:
+  if (visitor->enterForallStmt(this)) {
+    for_alist(next_ast, fIter)
+      next_ast->accept(visitor);
+    // not doing: fIter->accept(visitor);
+    fWith->acceptFI(visitor); // aka visitor->visitForallIntents(fWith);
+    fMBlock->accept(visitor); // includes loopBody
+    visitor->exitForallStmt(this);
+  }
+*/
+}
+
+Expr* ForallStmt::getFirstChild() {
+  return fIterVars.head;
+}
+
+Expr* ForallStmt::getFirstExpr() {
+  if (/*Expr* iv =*/ fIterVars.head->getFirstExpr())
+    INT_ASSERT(false); // todo: instead return iv;
+  else if (Expr* ie = fIterExprs.head->getFirstExpr())
+    return ie;
+
+  // we must have found something
+  INT_ASSERT(false);
+  return NULL; //dummy
+}
+
+Expr* ForallStmt::getNextExpr(Expr* expr) {
+  if (expr == fIterVars.tail)
+    return fIterExprs.head;
+
+  if (expr == fIterExprs.tail) {
+    if (Expr* inv = fIntentVars.head)
+      return inv;
+    else
+      return fLoopBody->getFirstExpr();
+  }
+
+  // Should we also descend into fWith?
```

Comment:
Right. #5295 works without descending into fWith. This is here as a reminder of sorts.